net/autossh/src/opnsense/mvc/app/controllers/VerbNetworks/Autossh/Api/KeysController.php

```diff
@@ -0,0 +1,219 @@
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+    ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+namespace VerbNetworks\Autossh\Api;
+
+use \OPNsense\Core\Backend;
+use \OPNsense\Core\Config;
+use \OPNsense\Base\ApiControllerBase;
+use \OPNsense\Base\UIModelGrid;
+use \VerbNetworks\Autossh\Autossh;
+
+class KeysController extends ApiControllerBase
+{
+    
+    public function searchAction()
+    {
+        $this->sessionClose(); // close out long running actions
+        $model = new Autossh();
+        $grid = new UIModelGrid($model->keys->key);
+        
+        $grid_data = $grid->fetchBindRequest(
+            $this->request,
+            array('name', 'type', 'key_fingerprint', 'timestamp'),
+            'name'
+        );
+        
+        foreach ($grid_data['rows'] as $row_index => $row_data) {
+            foreach ($row_data as $key => $value) {
+                if ($key === 'timestamp') {
+                    $grid_data['rows'][$row_index][$key] = date('Y-m-d H:i:s', $value);
+                } elseif ($key === 'key_fingerprint') {
+                    $fingerprint_elements = explode(' ', $value);
+                    $grid_data['rows'][$row_index][$key] = str_replace(
+                        'md5:',
+                        '',
+                        strtolower($fingerprint_elements[1])
+                    );
+                }
+            }
+        }
+        
+        return $grid_data;
+    }
+    
+    public function getAction($uuid = null)
+    {
+        $model = new Autossh();
+        if ($uuid != null) {
+            $node = $model->getNodeByReference('keys.key.'.$uuid);
+            if ($node != null) {
+                $data = $node->getNodes();
+                
+                // munge the data a little bit making it easier to use
+                $data['timestamp'] = date('Y-m-d H:i:s', $data['timestamp']);
+                $fingerprint_elements = explode(' ', $data['key_fingerprint']);
+                $data['key_fingerprint'] = str_replace('md5:', '', strtolower($fingerprint_elements[1]));
+                unset($data['key_private']);
+                
+                return array('key' => $data);
+            }
+        } else {
+            $node = $model->keys->key->add();
+            return array('key' => $node->getNodes());
+        }
+        return array();
+    }
+    
+    public function infoAction($uuid = null)
+    {
+        $ssh_key_restrictions = 'command="",no-agent-forwarding,no-pty,no-user-rc,no-X11-forwarding';
```

Comment:
This is listed in the volt, no idea if this is really necessary?

Reply:
yes - displaying the ssh public key to the user in a manner that is designed to prevent other security accidents is a deliberate thing as there will be some that may not understand the potential flow on effects - anyone that wants to strip off the key restrictions ought to know what they are doing

Reply:
Ok, looked a bit strange in UI bit I didn't have a deeper look into it yet.